Posts: Api integration -- Connect API for Post CRUD operations

BMariscal opened the pull request:
[Quick video demo](https://streamable.com/xxwzpp)
![](https://i.ibb.co/2cHyHyH/Screen-Shot-2020-06-18-at-6-07-19-PM.png)
------

 - On first entering the **/feed** page, the user can see the first 120 characters of every post.
- A not logged-in user is redirected to login/signup if they try to create a post.
- Any user can see the posts in the feed.
- Any user can create a share link to a post. 
- Only the post creator can edit or delete the post.
- The feed has pagination and infinite scrolling. 
-----
**Each Post:** 
 1. each post has a "View More" link which takes the user to the post's page at `/post/{postID}`.
 2. each post has a dropdown menu accessed by clicking on ". . . " that only the post's creator can see. 
 3. editing item in Feed brings author directly to editing.
4. editing item in post's page, takes author to Feed.
-----
**Feed:**
![](https://i.ibb.co/zR90N3R/Screen-Shot-2020-06-18-at-6-10-56-PM.png)
**Post Page:** `/post/{postID}`
1. Once in the post's page, the user can toggle the posts content to  "View More" or "View Less".
2. The post's creator has access to the dropdown menu with the "delete" and "edit" options.

    * Edit: 
      - A content pre-loaded modal pops up and allows the content creator to edit their post. 

    * Delete: 
      - A confirmation modal pops up to confirm whether the content creator wishes to delete the post.  If the user deletes their post in the post page, they'll be redirected back to the `/feed` page.

3. In the post's page, any user can see the social media icons to share the link, likes and see the comments. 
![](https://i.ibb.co/n7VNC0k/Screen-Shot-2020-06-18-at-6-11-49-PM.png)

![](https://i.ibb.co/Y8vJQ84/Screen-Shot-2020-06-18-at-6-11-34-PM.png)

![](https://i.ibb.co/txNQNmw/Screen-Shot-2020-06-18-at-6-11-18-PM.png)

-------
Tasks: 
A logged-in user can:

1. Create a Post 
   - [x] Make a `POST` request to `/api/posts/`, sending along the required data in the request body to the server.

2. Access any post (no auth required)
   - [x] Make a `GET` request to `/api/posts/:postId`.

3. Edit a post that they authored 
   - [x] Make a `PATCH` request to `/api/posts/:postId`, sending along the required data in the request body to the server.

4. Delete a post that they authored
   - [x] Make a `DELETE` request to `/api/posts/:postId`.  
-------
Note: tasks 5 and 6 will be moved to a different PR: 
5. Load Comments
    - [ ] Load the comments!
         https://github.com/FightPandemics/FightPandemics/issues/550
6. Like & Unliking Comment
    - [ ] https://github.com/FightPandemics/FightPandemics/issues/183


More references:
https://www.notion.so/fightpandemics/Feed-653658508c4649869787c48f3e4eee13


Closes #182
Closes #698

robinv85 commented on client/src/components/CreatePost/Form/Form.js:
```diff
@@ -109,26 +143,36 @@ const ModalComponent = ({ setCurrentStep, onClose }) => {
           <AskHelp
             formData={formData}
-            handleSubmit={handleSubmit}
+            handleSubmit={(e) => handleSubmit(e)}
```
any reason why you add the `(e) =>` arrow function? I prefer to just use the handler directly, with hooks it doesn't make a difference but with class components you avoid recreating the handler functions for every render

BMariscal replied:
just my preference but i don't really mind changing it back since it'll match the rest of the code. I changed it back!

robinv85 commented on client/src/pages/Feed.js:
```diff
@@ -242,7 +242,12 @@ const Feed = (props) => {
+    if (isAuthenticated) {
+      dispatchAction(TOGGLE_STATE, "createPostModal");
+    } else {
+      history.push(LOGIN);
```
I think this replaces the #505 pull request
because this is using redirect from the code instead of a wrapper component
probably works better because this is a modal?

BMariscal replied:
Hmm, the redirect was not working for me before this change.

idjevm replied:
I also agree with @robinv85, I think we should close #505 PR for now, since once this will be merged, it would take care of the redirecting if the user is not authenticated...

robinv85 replied:
Yeah @BMariscal I'm not saying this was working, I mean that there was probably a bug on github, I just saw someone else opened a PR too for that #668 which is the same code as here pretty much.
So I think this solution probably works better at first sight than the other in #505 so we can probably close the other ones.

silenceway replied:
https://github.com/FightPandemics/FightPandemics/pull/668 covers when you navigate to the create-post page, this only covers when you open the create post modal.

idjevm replied:
@silenceway Isn't #668 is exactly the same as this part? The only difference that I noticed was that the if/else are inverted. Not sure if I missed anything big, if so please let me know so I can think about that too.

silenceway replied:
According to the design, create a post on mobile is a full page and the desktop page is modal. The design team should review that.

BMariscal commented on client/src/hooks/reducers/feedReducers.js:
```diff
@@ -74,18 +92,18 @@ export const postsReducer = (state = postsState, action) => {
+        posts: {
+          ...state.posts,
+          [action.postId]: {
+            ...state.posts[action.postId],
```
Auto formatted by linter

idjevm commented on client/package-lock.json:
```diff
@@ -15325,6 +15325,22 @@
         }
       }
     },
+    "react-contenteditable": {
```
@robinv85 Do we need to add this file on this PR? I am not sure, I see the package.json file being updated so perhaps yes? @BMariscal I did not see any usages of this package, but maybe I am missing that part.

robinv85 replied:
yes there is a new dependency    "react-contenteditable": "^3.3.4",

idjevm commented on client/src/components/CreatePost/Form/Form.js:
```diff
@@ -58,16 +62,30 @@ const ModalComponent = ({ setCurrentStep, onClose }) => {
+    payload["types"] = formData.tags;
+    payload["visibility"] = formData.shareWith;
+
+    payload["expireAt"] = payload["expireAt"].toLowerCase();
```
Is it not possible to do `formData.expires.toLowerCase()` and `formData.shareWith.toLowerCase()`? 

I think it should allow you to do that since `expires` and `shareWith` seem to be strings...

robinv85 replied:
yeah here it's better to use dot notation
bracket notation is usually used for dynamic keys `obj[variableName]` or used to be more for characters that aren't allowed in JS syntax but nowadays with modern ECMASCript those are very rare cases I think, maybe for example built-in keywords such as `"class"`

BMariscal commented:
Hello @robinv85 
I can incorporate #550 in this. No worries.

You should wait until this gets merged before working on #183

Naraujo13 commented on backend/lib/endpoints/posts.js:
```diff
@@ -321,7 +327,7 @@ async function routes(app) {
         throw app.httpErrors.internalServerError();
       } else if (post === null) {
         throw app.httpErrors.notFound();
-      } else if (!userId.equals(post.author.id)) {
+      } else if (userId !== post.author.id.toString()) {
```
Why change this to use string comparison instead of the ObjectId equals?

BMariscal replied:
I had to change it to this because I wasn't getting the signup email to validate my account. It was an issue with my .env configs.

I'll change it back!  Thanks for bringing this to my attention. This is still wip.

Naraujo13 commented on backend/lib/endpoints/posts.js:
```diff
@@ -104,20 +104,18 @@ async function routes(app) {
 
       // Unlogged user limitation for post content size
       /* eslint-disable sort-keys */
-      const contentProjection = user
-        ? "$content"
-        : {
```
Removing this conditional will cause all posts to be capped to 150 char even if user is logged. Is it intentional?

BMariscal replied:
So this is the post content that's retrieved in the feed. There is no toggling for more content in the feed, this now happens in the post page in the `posts/{postID}` route. I just need the first 120 characters of the post's content for every post in the feed for everyone, logged in or not. This is the intended behavior.

Naraujo13 replied:
Ok, this seems okay!

Naraujo13 commented on backend/lib/endpoints/posts.js:
```diff
@@ -258,46 +257,53 @@ async function routes(app) {
-            $match: {
-              parentId: null,
-              postId: mongoose.Types.ObjectId(postId),
+      const [comment] = await app.to(Comment.findOne({ postId }));
```
Why this query before getting the comments? Is it to check if there are any comments at this post?

Naraujo13 replied:
Okay, this will work.
There is probably a way of doing this check at the same query to avoid 2 requests to the DB, but I will open a ticket and look into that, later on, to not block this now

Naraujo13 commented on backend/lib/models/Post.js:
```diff
@@ -9,13 +9,13 @@ const POST_TYPES = [
   "Entertainment",
   "Funding",
-  "Groceries/Food",
+  "Groceries / Food",
```
Why adding these spaces with slashes at categories?
Just to know because if we follow this need to update the data model docs also

BMariscal replied:
I changed this so it would match the frontend post-type value to allow PATCH/edits to a post. If it is going to be too much a hassle to update the docs, I can just change this property value in the frontend or do a mapping, no worries.

joshmorel replied:
You'll need to merge/rebase on master to resolve conflicts but anyways and you will see we recently updated filterOptions to match backend model - https://github.com/FightPandemics/FightPandemics/blob/master/client/src/assets/data/filterOptions.js. 

Ideally we'd have a shared/common constants like below. Might be a good post-MVP project to initiate? Keep model and as-is and update front-end options to match model.

```javascript
const POST_TYPES = {
INFORMATION: "Information",
LEGAL: "Legal",
}
```
```

Naraujo13 replied:
It is easy to update it at documentation, just have to watch out to keep it consistent across the app.
That seems nice @joshmorel

Naraujo13 replied:
Updating docs is easy, just have to watch out to keep it consistent across the app.
Also, that sounds good @joshmorel

BMariscal replied:
I'll do the mapping in the frontend. Don't change the doc @Naraujo13 The mapping should be super easy since I know which fields to target. Thank you both @Naraujo13 @joshmorel ❤️

joshmorel commented on backend/lib/endpoints/posts.js:
```diff
@@ -480,10 +486,10 @@ async function routes(app) {
           },
           {
-            $limit: limit || COMMENT_PAGE_SIZE,
+            $limit: parseInt(limit) || COMMENT_PAGE_SIZE,
```
Weird the `getCommentsSchema` should be validating the querystring and transforming these to ints - but it's doing nothing at all (where as `getPostsSchema` works fine)`. Something to look into later I guess because this works.

Naraujo13 replied:
Yeah, I found that while testing also. Not sure why the schema validation seems to be different between those two endpoints.

joshmorel commented on client/src/assets/data/filterOptions.js:
```diff
@@ -21,11 +21,11 @@ export default {
     options: [
       "Medical Supplies",
-      "Groceries/Food",
+      "Groceries / Food",
```
Or maybe you did merge master but update the change made in 53a3583a987773eabb2216854cec8f0200d75558 ?

Anyways, let's keep no spaces in both f/e & model (no mapping).

BMariscal replied:
Sounds good to me, as long as there is consistency. I'll change it to no spaces. Thanks Josh!

idjevm commented on client/src/components/CreatePost/Form/AskHelp.js:
```diff
@@ -3,15 +3,24 @@ import First from "./FirstSection";
 import { Footer, Submit } from "components/CreatePost/StyledModal";
+import createPostSettings from "assets/data/createPostSettings";
+
+const { helpTypes } = createPostSettings;
```
Uhhh Ohh :O this file was deleted recently....

BMariscal replied:
Thanks!!

idjevm commented on client/src/components/CreatePost/Form/EditForm.js:
```diff
@@ -0,0 +1,175 @@
+import { Divider, ModalWrapper } from "components/CreatePost/StyledModal";
+import OfferHelp from "components/CreatePost/Form/OfferHelp";
+import AskHelp from "components/CreatePost/Form/AskHelp";
+import { postToFormData, formDataToPostPatch } from "constants/formToPostMappings";
+
+const { TabPane } = Tabs;
+
+const errorMsg = {
+  title: "Please include a title for your post.",
+  description: "Please include a description for your post.",
+  help: "Please select a type of help.",
+  tags: "Please add at least one tag.",
+};
+
```
This file was also deleted or modified I believe

idjevm commented on client/src/components/CreatePost/Form/OfferHelp.js:
```diff
@@ -3,15 +3,25 @@ import First from "./FirstSection";
+import createPostSettings from "assets/data/createPostSettings";
+
+const { helpTypes } = createPostSettings;
 
```
This file was also deleted so probably gonna be a lot of merge conflicts 😨

BMariscal replied:
I'm working on fixing the merge conflict rn. Thanks for letting me know the PR that led to these conflicts @idjevm that was super helpful!  if you see anyone making changes to the post/feed please lmk

joshmorel replied:
Everything looks good except I believe through the merge resolution the two unused form components were added back - AskHelp.js & OfferHelp.js - I will approve so you can merge right after deleting those two files first:

 client/src/components/CreatePost/Form/AskHelp.js
 client/src/components/CreatePost/Form/OfferHelp.js

BMariscal commented:
@joshmorel  I just fixed the merge conflicts and addressed the changes requested.
I added this feature: 
`In Feed post menu has only delete; while in PostPage it has both edit/delete; I think it would be ideal to have edit item in Feed to bring author directly to editing`

The liking/commenting is still not integrated, so this issue `The liking works on the Feed, but not PostPage` will be addressed in a different PR.

idjevm commented on client/src/pages/Feed.js:
```diff
@@ -406,15 +423,49 @@ const Feed = (props) => {
+        }
+      } catch (error) {
+        console.log({
+          error,
```
weird formatting from ESLINT probably? or is it this one the right way? I saw others above that are different than this one.

BMariscal replied:
It was the linter. Yeah, Robin mentioned the linter was doing weird things lately, idk if he figured out why.

joshmorel commented:
Thanks @BMariscal I'll go ahead and merge. We should split out liking/unliking posts & commenting to two different engs/PRs.
